PUB-7961 Extended instrumentor metrics

lemonion opened the pull request:
Add more metrics to instrumentor library. Will be used in outbox to capture network requests.

lemonion committed:
extended instrumentor metrics

tmseiler commented on instrumentor-core/src/main/java/com/sproutsocial/metrics/Instrumentor.java:
```diff
@@ -31,11 +26,15 @@
         private Counter inFlight;
 
         Context(String name) {
+            totalMeter = metricRegistry.meter(name + ".total");
```
`total` is slightly ambiguous, maybe `total_started`? not sure on the naming convention

lemonion replied:
I've seen a mix of camel and snake case around. Changed to `totalStarted` to be consistent with `inFlight`.

maxrjohnson23 replied:
Do we need this metric?  It seems like we already have total method calls with this meter: `com.mycompany.Example.sayHello.count -- a lifetime count of calls to the method`

We could find the success rate using total count - error count as well, and we also have an error rate available.  I'm not sure what we're trying to measure.

lemonion replied:
Thanks! Somehow I missed that Timer is also a meter and can report the total number of calls. I have removed the total meter. 
I do believe we should keep the success rate instead of implying success rate being the difference of total and error in order to make the success meter directly comparable to the error meter.

tmseiler commented on instrumentor-core/src/main/java/com/sproutsocial/metrics/Instrumentor.java:
```diff
@@ -109,9 +108,12 @@ public MetricRegistry getMetricRegistry() {
             context.inFlight.inc();
             try (@SuppressWarnings("unused") Timer.Context ctx = context.timer.time()){
-                return callable.call();
+                T result = callable.call();
```
If this still works, I think this is slightly clearer/safer, to avoid someone accidentally inlining `result` into that `return`. It will also minimize the amount of things in the `try`

```
            T result;
            try (@SuppressWarnings("unused") Timer.Context ctx = context.timer.time()){
                result = callable.call();
            } catch (Exception e) {
                if (exceptionFilter.test(e)) {
                    context.errorMeter.mark();
                }
                throw e;
            } finally {
                context.inFlight.dec();
            }             
            context.successMeter.mark();
            return result;
```

tmseiler commented on instrumentor-core/src/main/java/com/sproutsocial/metrics/Instrumentor.java:
```diff
@@ -31,11 +26,15 @@
 
 
     private class Context {
```
I know this isn't your problem to solve right now, but it confused me that there's this competing `Context` class used immediately next to the one from `Metrics` package

lemonion replied:
Some ideas:
- Rename this to `InstrumentorContext` or `InstrumentationContext` and keep it where it is. -> `InstrumentorContext` inside a `Instrumentor` class seems redundant (in naming), but might help with naming conflict confusion.
- Rename this to `InstrumentorContext` and move this to top-level class. -> Doesn't seem necessary to expose this internal class as part of the package.

Happy to hear about your thoughts as well.

tmseiler replied:
`InstrumentationContext` would at least be consistent with the factory method `createInstrumentationContext`. Maybe we can use `timerContext` as the name in the `try-with-resources`

tmseiler commented:
Looks pretty good, just left some clarity suggestions

lemonion committed:
improve success logging

lemonion committed:
change total meter name

jsadn commented:
i'm curious as to the need to modify this...we should already be capturing each network request and response in our logs for facebook, instagram and twitter. 

my vote would also be to stop using this Instrumentor thing and improve observability patterns within the outbox code itself.

jsadn commented:
to expand on my comment above - I think we should move our observability code to use a [DomainProbe](https://martinfowler.com/articles/domain-oriented-observability.html) like we've been using in airlock - rather than this observer thing and lambdas, which are hard to debug.

micahben reviewed (commented):
I have the [same question as Max](https://github.com/sproutsocial/instrumentor/pull/28#discussion_r735682098).

If you _do_ add this, you should update the README.

FWIW, I'm not a huge advocate of this library. It introduces new annotations, when it would be much better to support the existing dropwizard-metrics annotations, and build upon those. A library like https://github.com/palominolabs/metrics-guice does just that.

lemonion committed:
improve context naming

lemonion committed:
remove total meter

lemonion committed:
update readme

lemonion commented:
To reviewers:
- I have removed the total count meter and kept the success meter based on Max's feedback. I had also updated the README according to Micah's feedback.
- Yes, we do have instrumentation around some existing network calls in outbox, but not all. The idea here is to capture all the network calls downstream of the publishers so we have an idea of what's inflight, how many has started, succeeded, failed and how long those network calls are taking. This will help pus in troubleshooting whether the issue is in our code or downstream network. 
- We originally tried to spin our own metrics in outbox directly, but soon arrived at a similar pattern as this instrumentor library. The motivation is to allow a small change to capture all the metrics we care about, instead of adding them individually.
- Since this library is not widely used, the change (adding success meter) should be pretty low risk. The plan is to test it out in the publishers and hopefully be useful for other projects as well.

micahben reviewed (dismissed):
You may or may not be aware that this is a publicly available repository, and that we publish artifacts to OSS Sonatype. I haven't cut a new version of this in a long time, but the guidance for doing so can generally be [found here](https://central.sonatype.org/publish/publish-guide/). I'm happy to try to help if you have any questions.

lemonion committed:
update AUTHORS and developer section for sonatype
